src/main/java/org/prebid/server/bidder/readpeak/ReadPeakBidder.java

```diff
@@ -0,0 +1,204 @@
+        } else if (request.getApp() != null) {
+            final App app = request.getApp().toBuilder()
+                    .id(StringUtils.isNotBlank(extImpReadPeak.getSiteId())
+                            ? extImpReadPeak.getSiteId() : request.getApp().getId())
+                    .publisher(publisher)
+                    .build();
+            requestBuilder.app(app);
+        }
+
+        final BidRequest outgoingRequest = requestBuilder.build();
+
+        return BidderUtil.defaultRequest(outgoingRequest, endpointUrl, mapper);
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(BidderCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            final List<BidderBid> bids = extractBids(bidResponse);
+            return Result.withValues(bids);
+        } catch (DecodeException | PreBidException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private List<BidderBid> extractBids(BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return Collections.emptyList();
+        }
+        return bidsFromResponse(bidResponse);
+    }
+
+    private List<BidderBid> bidsFromResponse(BidResponse bidResponse) {
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(List::stream)
+                .map(bid -> makeBid(bid, bidResponse))
+                .filter(Objects::nonNull)
+                .toList();
+    }
+
+    private BidderBid makeBid(Bid bid, BidResponse bidResponse) {
+        final Bid resolvedBid = resolveMacros(bid);
+        final BidType bidType = getBidType(bid);
+        final Bid updatedBid = addBidMeta(resolvedBid);
+        return BidderBid.of(updatedBid, bidType, bidResponse.getCur());
+    }
+
+    private Bid resolveMacros(Bid bid) {
+        final BigDecimal price = bid.getPrice();
+        final String priceAsString = price != null ? price.toPlainString() : "0";
+
+        return bid.toBuilder()
+                .nurl(StringUtils.replace(bid.getNurl(), PRICE_MACRO, priceAsString))
+                .adm(StringUtils.replace(bid.getAdm(), PRICE_MACRO, priceAsString))
+                .burl(StringUtils.replace(bid.getBurl(), PRICE_MACRO, priceAsString))
+                .build();
+    }
+
+    private BidType getBidType(Bid bid) {
+        final Integer markupType = ObjectUtils.defaultIfNull(bid.getMtype(), 0);
+
+        return switch (markupType) {
+            case 1 -> BidType.banner;
+            case 4 -> BidType.xNative;
+            default -> throw new PreBidException(
+                    "Unable to fetch mediaType " + bid.getMtype() + " in multi-format: " + bid.getImpid());
+        };
```

Comment:
Errors should be collected ignoring only invalid bids, and the code should fail on the very first invalid bid